src/Sulu/Bundle/DocumentManagerBundle/Reference/Provider/DocumentReferenceProvider.php

```diff
@@ -0,0 +1,131 @@
+use Sulu\Component\Content\ContentTypeManagerInterface;
+use Sulu\Component\Content\Document\Behavior\SecurityBehavior;
+use Sulu\Component\Content\Document\Behavior\StructureBehavior;
+use Sulu\Component\Content\Document\Behavior\WorkflowStageBehavior;
+use Sulu\Component\DocumentManager\Behavior\Mapping\TitleBehavior;
+use Sulu\Component\DocumentManager\Behavior\Mapping\UuidBehavior;
+
+class DocumentReferenceProvider implements DocumentReferenceProviderInterface
+{
+    protected ContentTypeManagerInterface $contentTypeManager;
+
+    protected StructureManagerInterface $structureManager;
+
+    protected ReferenceRepositoryInterface $referenceRepository;
+
+    protected DocumentInspector $documentInspector;
+
+    protected string $structureType;
+
+    protected string $referenceSecurityContext;
+
+    public function __construct(
+        ContentTypeManagerInterface $contentTypeManager,
+        StructureManagerInterface $structureManager,
+        ReferenceRepositoryInterface $referenceRepository,
+        DocumentInspector $documentInspector,
+        string $structureType,
+        string $referenceSecurityContext
+    ) {
+        $this->contentTypeManager = $contentTypeManager;
+        $this->structureManager = $structureManager;
+        $this->referenceRepository = $referenceRepository;
+        $this->documentInspector = $documentInspector;
+        $this->structureType = $structureType;
+        $this->referenceSecurityContext = $referenceSecurityContext;
+    }
+
+    public function updateReferences(UuidBehavior&TitleBehavior&StructureBehavior $document, string $locale): ReferenceCollector
+    {
+        $referenceResourceKey = $this->getReferenceResourceKey($document);
+
+        if (!$referenceResourceKey) {
+            throw new \Exception('ReferenceResourceKey must be defined');
+        }
+
+        $workflowStage = $document instanceof WorkflowStageBehavior ? $document->getWorkflowStage() : 0;
+
+        $referenceCollector = new ReferenceCollector(
+            $this->referenceRepository,
+            $referenceResourceKey,
+            $document->getUuid(),
+            $locale,
+            $document->getTitle(),
+            $this->getReferenceSecurityContext($document),
+            $document->getUuid(),
+            $this->getReferenceSecurityObjectType(),
+            $workflowStage
+        );
+
+        $structure = $document->getStructure();
+        $templateStructure = $this->structureManager->getStructure($document->getStructureType(), $this->getStructureType());
+        foreach ($templateStructure->getProperties(true) as $property) {
+            $contentType = $this->contentTypeManager->get($property->getContentTypeName());
+
+            if (!$contentType instanceof ReferenceContentTypeInterface) {
+                continue;
+            }
+
+            $contentType->getReferences($property, $structure->getProperty($property->getName()), $referenceCollector);
```

Comment:
Maybe we can call here:  `$property->setValue($structure->getProperty($property->getName()))`